src/mrinufft/operators/off_resonance.py

```diff
@@ -0,0 +1,330 @@
+        Also supports Cupy arrays and Torch tensors.
+        Expected shape is ``(nz, ny, nx)``.
+    readout_time : np.ndarray
+        Readout time in ``[s]`` of shape ``(nshots, npts)`` or ``(nshots * npts,)``.
+        Also supports Cupy arrays and Torch tensors.
+    n_time_segments : int, optional
+        Number of time segments. The default is ``6``.
+    n_bins : int | Sequence[int] optional
+        Number of histogram bins to use for ``(B0, T2*)``. The default is ``(40, 10)``
+        If it is a scalar, assume ``n_bins = (n_bins, 10)``.
+        For real fieldmap (B0 only), ``n_bins[1]`` is ignored.
+    mask : np.ndarray, optional
+        Boolean mask to avoid histogram of background values.
+        The default is ``None`` (use the whole map).
+        Also supports Cupy arrays and Torch tensors.
+
+    Returns
+    -------
+    B : np.ndarray
+        Temporal interpolator of shape ``(n_time_segments, len(t))``.
+        Array module is the same as input fieldmap.
+    C : np.ndarray
+        Off-resonance phase map at each time segment center of shape
+        ``(n_time_segments, *fieldmap.shape)``.
+        Array module is the same as input fieldmap.
+    """
+    # default
+    if isinstance(n_bins, (list, tuple)) is False:
+        n_bins = (n_bins, 10)
+
+    # transform to list
+    n_bins = list(n_bins)
+
+    # get backend and device
+    xp = get_array_module(fieldmap)
+
+    # enforce complex field
+    fieldmap = xp.asarray(fieldmap, dtype=xp.complex64)
+
+    # cast arrays to fieldmap backend
+    if xp.__name__ == "torch":
+        is_torch = True
+    else:
+        is_torch = False
+
+    if is_cuda_array(fieldmap):
+        assert CUPY_AVAILABLE, "GPU computation requires Cupy!"
+        xp = cp
+        fieldmap = _to_cupy(fieldmap)
+        readout_time = _to_cupy(readout_time)
+        mask = _to_cupy(mask)
+    else:
+        xp = np
+        fieldmap = _to_numpy(fieldmap)
+        readout_time = _to_numpy(readout_time)
+        mask = _to_numpy(mask)
+
+    readout_time = xp.asarray(readout_time, dtype=xp.float32).ravel()
+    if mask is None:
+        mask = xp.ones_like(fieldmap, dtype=bool)
+    else:
+        mask = xp.asarray(mask, dtype=bool)
+
+    # get field map
+    if xp.isreal(fieldmap).all().item():
+        r2star = None
+        b0 = fieldmap
+        fieldmap = 0.0 + 1j * b0
+    else:
+        r2star = fieldmap.real
+        b0 = fieldmap.imag
+
+    # Hz to radians / s
+    fieldmap = 2 * math.pi * fieldmap
+
+    # create histograms
+    if r2star is not None:
+        z = fieldmap[mask].ravel()
+        z = xp.stack((z.imag, z.real), axis=1)
+        hk, ze = xp.histogramdd(z, bins=n_bins)
+        ze = list(ze)
+
+        # get bin centers
+        zc = [e[1:] - (e[1] - e[0]) / 2 for e in ze]
+
+        # complexify
+        zk = _outer_sum(1j * zc[0], zc[1])  # [K1 K2]
+        zk = zk.T
+        hk = hk.T
+    else:
+        z = fieldmap[mask].ravel()
```

Comment:
This line can be factorized out of the if statement

Reply:
Good idea!